### PR TITLE
Fix compiler warnings and treat warnings as errors

### LINKED ISSUE #779: Fix All Compiler Warnings
### Description of the task
With our migration to bazel, we have started to get a _lot_ of compiler warnings that we just didn't have turned on before. We should fix them all and make warnings fail compilation.

### Acceptance criteria

- [ ] Make warnings errors (so they will fail compilation)
- [ ] Fix all warnings in the codebase

### PR DESCRIPTION
**Pull request by scveloso**
### Description
- Fix all compiler warnings
- "Treat warnings as errors" setting added in bazelrc

### Testing Done
- Bazel build should fail when encountering any warnings
- Old unit tests still pass

### Resolved Issues
resolves #779 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### Misc
So for some of the compiler warnings, I had to make some decisions that I'm not sure are 100% correct:

#### "Comparing signed int against unsigned int" warning in for-loops
A common error was when we would do ```for (int i = 0; i < collection.size() ...)```. Since any size() is an unsigned int, I needed to change it to ```for (auto i = 0U; i < collection.size() ...)``` to make it unsigned. I can change it to something else but I felt this is the most similar to our typical for-loops (I could be wrong). 

#### "X will be initialized after Y" warning
Just a matter of reordering the definition of members in a class so that they match with the order that they are initialized in the implementation of the constructor (or vice-versa). Commonly found in the *_tactic classes for example ```GoalieTactic(...) : ball(ball), field(field), ..., Tactic(...)``` had to be changed to ```GoalieTactic(...) : Tactic(...), ball(ball), field(field), ...```

#### "Variable unused" warning
A couple of unused variables that I removed. I did a quick check that there isn't any function or behavior happening when these variables are initialized. Especially some of the std::pair ones where only the first part of the pair was being used. So, for example, I changed ```auto [first, second] = pair``` to ```auto first = pair.first``` if second wasn't being used.

### CODE REVIEW

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (scveloso) [inline src/software/ai/hl/stp/tactic/goalie_tactic.cpp:99]: Can someone who knows this better than I do verify that this is OK?

File: src/software/ai/hl/stp/tactic/crease_defender_tactic.cpp
Comment (scveloso) [inline src/software/ai/hl/stp/tactic/crease_defender_tactic.cpp:184]: Can someone who knows this better than I do verify that this is OK?

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (scveloso) [inline src/software/ai/hl/stp/tactic/goalie_tactic.cpp:129]: Can someone who knows this better than I do verify that this is OK?

File: src/software/ai/hl/stp/tactic/receiver_tactic.cpp
Comment (scveloso) [inline src/software/ai/hl/stp/tactic/receiver_tactic.cpp]: This variable was unused, can I remove this?

**Review by henryymliu — CHANGES_REQUESTED** [review]
Unsigned int comparisons: I generally only prefer using auto if the type is very long, this use case is definitely not one of them. Consider replacing `auto` with `unsigned` or `size_t`, and then you can drop the `U` suffix on `0U`.

Also, all the move_action arg changes and everything else LGTM

**Comment by scveloso** [comment]
> Unsigned int comparisons: I generally only prefer using auto if the type is very long, this use case is definitely not one of them. Consider replacing `auto` with `unsigned` or `size_t`, and then you can drop the `U` suffix on `0U`.
> 
> Also, all the move_action arg changes and everything else LGTM

Thanks for the review, updated to use `unsigned` instead of `auto`.

File: src/software/ai/hl/stp/tactic/receiver_tactic.cpp
Comment (henryymliu) [reply src/software/ai/hl/stp/tactic/receiver_tactic.cpp]: Oops didn't see this comment, @garethellis0 @MathewMacDougall ?

Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/receiver_tactic.cpp]: The commented-out one directly below this comment? Yeah, go ahead, thanks for checking!

**Review by EvanMorcom — COMMENTED** [review]
LGTM :+1: 

There are a lot of changes that are just moving variable declarations around.

Ideally, if these don't serve a purpose it would be nice to revert them to reduce the files changed.

File: src/software/ai/ai.h
Comment (scveloso) [reply src/software/ai/ai.h:45]: Yeah, any changes where I reordered variable declarations are because of a bazel warning where the order of variable declarations don't match the order in which they are initialized in the .cpp implementation.

File: src/software/geom/util.cpp
Comment (scveloso) [reply src/software/geom/util.cpp:701]: These were causing a bazel compiler warning since it was unused. Since compiler warnings are now treated as errors, this will cause any build to fail. I could comment it out maybe?

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (scveloso) [reply src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: Fixed in the update

**Comment by scveloso** [comment]
Thanks for the reviews. It seems like the drop in coverage is due to unused code being removed (I might be wrong). I've addressed all comments, let me know if there's anything else!

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada: - just see my one comment before merging please

**Comment by scveloso** [comment]
None of the unused code removals were being tested. Waiting on @henryymliu requested changes resolution I think, then I can merge.

**Comment by MathewMacDougall** [comment]
@scveloso this is ready to merge pending @henryymliu ?

**Review by henryymliu — DISMISSED** [review]
LGTM, sorry for the delay

**Comment by garethellis0** [comment]
@scveloso  - looks like there are some merge conflicts here

**Comment by scveloso** [comment]
@garethellis0 @MathewMacDougall Resolved merge conflicts from recently updated master. Sorry looks like resolving them auto-dismissed @henryymliu 's review. Hoping to get two approvals soon - merge conflicts are going to be common since this PR touches a lot of the codebase.

Edit: ~~the build may fail since there are some issues caused by resolving merge conflicts. Am fixing them and will update this comment when they're fixed.~~ Now fixed and conflicts are resolved

**Review by MathewMacDougall — APPROVED** [review]
All good, let's try get this in quick :zap:

File: src/software/ai/ai.h
Comment (EvanMorcom) [reply src/software/ai/ai.h:45]: Sweet, didn't know that 👍

File: src/software/geom/util.cpp
Comment (EvanMorcom) [reply src/software/geom/util.cpp:701]: I'll agree with Gareth here, let's just delete it for now.

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp
Comment (EvanMorcom) [reply src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: 👍

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada: - thanks for all the work on this!

**Pull request merged**
